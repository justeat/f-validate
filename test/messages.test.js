import TestUtils from 'js-test-buddy';
import FormValidation from '../src';

describe('error messages', () => {

    describe('custom', () => {

        it('should apply error class to invalid field', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required data-required-error="custom required error message" />
                                </form>`);
            const form = document.querySelector('form');
            const validateForm = new FormValidation(form);

            // Act
            validateForm.isValid();

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should not apply error class to valid field', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required value="x" data-required-error="custom required error message" />
                                </form>`);
            const form = document.querySelector('form');
            const validateForm = new FormValidation(form);

            // Act
            validateForm.isValid();

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should replace existing error message with new', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input 
                                        required 
                                        maxlength="3" 
                                        data-required-error="custom required error message"
                                        data-maxlength-error="custom maxlength error message"
                                    /> 
                                </form>`);
            const form = document.querySelector('form');
            const input = form.querySelector('input');
            const validateForm = new FormValidation(form);

            // Act & Assert
            validateForm.isValid();
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Make input invalid due to maxlength exceeded
            input.value = 'xxxx';

            validateForm.isValid();
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should hide error message if field is now valid', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                        <input required data-required-error="custom required error message" />
                                    </form>`);
            const form = document.querySelector('form');
            const input = form.querySelector('input');
            const validateForm = new FormValidation(form);

            // Act & Assert
            validateForm.isValid();
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Make input valid
            input.value = 'x';

            validateForm.isValid();
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

        });

    });

    describe('default', () => {

        it('should apply error class to invalid field', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required />
                                </form>`);
            const form = document.querySelector('form');
            const validateForm = new FormValidation(form);

            // Act
            validateForm.isValid();

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should not apply error class to valid field', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required value="x" />
                                </form>`);
            const form = document.querySelector('form');
            const validateForm = new FormValidation(form);

            // Act
            validateForm.isValid();

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should replace existing error message with new', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required maxlength="3" />
                                </form>`);
            const form = document.querySelector('form');
            const input = form.querySelector('input');
            const validateForm = new FormValidation(form);

            // Act & Assert
            validateForm.isValid();
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Make input invalid due to maxlength exceeded
            input.value = 'xxxx';

            validateForm.isValid();
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should hide error message if field is now valid', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                        <input required />
                                    </form>`);
            const form = document.querySelector('form');
            const input = form.querySelector('input');
            const validateForm = new FormValidation(form);

            // Act & Assert
            validateForm.isValid();
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Make input valid
            input.value = 'x';

            validateForm.isValid();
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

    });

    describe('grouped', () => {

        it('should display error messages grouped at the bottom (default)', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required />
                                    <input required minlength="2" value="x" />
                                </form>`);
            const form = document.querySelector('form', {
                groupErrors: true
            });
            const validateForm = new FormValidation(form);

            // Act
            validateForm.isValid();

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should display error messages grouped at the top', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required />
                                    <input required minlength="2" value="x" />
                                </form>`);
            const form = document.querySelector('form', {
                groupErrors: true,
                errorPlacementTopOfForm: true
            });
            const validateForm = new FormValidation(form);

            // Act
            validateForm.isValid();

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should not display error messages on valid form', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required value="x" />
                                    <input required minlength="2" value="xx" />
                                </form>`);
            const form = document.querySelector('form', {
                groupErrors: true
            });
            const validateForm = new FormValidation(form);

            // Act
            validateForm.isValid();

            // Assert
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should replace existing group error messages', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required />
                                    <input required minlength="2" />
                                </form>`);
            const form = document.querySelector('form', {
                groupErrors: true
            });
            const inputs = form.querySelectorAll('input');
            const validateForm = new FormValidation(form);

            // Act & Assert
            validateForm.isValid();
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Make input invalid due to maxlength exceeded
            inputs[0].value = 'x';
            inputs[1].value = 'x';

            validateForm.isValid();
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

        it('should hide existing group error message if group is now valid', () => {

            // Arrange
            TestUtils.setBodyHtml(`<form>
                                    <input required />
                                    <input required minlength="2" />
                                </form>`);
            const form = document.querySelector('form', {
                groupErrors: true
            });
            const inputs = form.querySelectorAll('input');
            const validateForm = new FormValidation(form);

            // Act & Assert
            validateForm.isValid();
            expect(TestUtils.getBodyHtml()).toMatchSnapshot();

            // Make inputs valid
            inputs[0].value = 'x';
            inputs[1].value = 'xx';

            validateForm.isValid();
            const html = TestUtils.getBodyHtml();
            expect(html).toMatchSnapshot();

        });

    });
});
