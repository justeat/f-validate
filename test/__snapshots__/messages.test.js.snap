// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error messages custom should apply error class to invalid field 1`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" data-required-error=\\"custom required error message\\" class=\\"has-error\\"><p class=\\"form-error\\">custom required error message</p>
                                </form>"
`;

exports[`error messages custom should hide error message if field is now valid 1`] = `
"<form class=\\"has-error\\">
                                        <input required=\\"\\" data-required-error=\\"custom required error message\\" class=\\"has-error\\"><p class=\\"form-error\\">custom required error message</p>
                                    </form>"
`;

exports[`error messages custom should hide error message if field is now valid 2`] = `
"<form class=\\"has-success\\">
                                        <input required=\\"\\" data-required-error=\\"custom required error message\\" class=\\"has-success\\"><p class=\\"form-error is-hidden\\"></p>
                                    </form>"
`;

exports[`error messages custom should not apply error class to valid field 1`] = `
"<form class=\\"has-success\\">
                                    <input required=\\"\\" value=\\"x\\" data-required-error=\\"custom required error message\\" class=\\"has-success\\">
                                </form>"
`;

exports[`error messages custom should replace existing error message with new 1`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" maxlength=\\"3\\" data-required-error=\\"custom required error message\\" data-maxlength-error=\\"custom maxlength error message\\" class=\\"has-error\\"><p class=\\"form-error\\">custom maxlength error message</p>
                                </form>"
`;

exports[`error messages custom should replace existing error message with new 2`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" maxlength=\\"3\\" data-required-error=\\"custom required error message\\" data-maxlength-error=\\"custom maxlength error message\\" class=\\"has-error\\"><p class=\\"form-error\\">custom maxlength error message</p>
                                </form>"
`;

exports[`error messages default should apply error class to invalid field 1`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" class=\\"has-error\\"><p class=\\"form-error\\">This field is required.</p>
                                </form>"
`;

exports[`error messages default should hide error message if field is now valid 1`] = `
"<form class=\\"has-error\\">
                                        <input required=\\"\\" class=\\"has-error\\"><p class=\\"form-error\\">This field is required.</p>
                                    </form>"
`;

exports[`error messages default should hide error message if field is now valid 2`] = `
"<form class=\\"has-success\\">
                                        <input required=\\"\\" class=\\"has-success\\"><p class=\\"form-error is-hidden\\"></p>
                                    </form>"
`;

exports[`error messages default should not apply error class to valid field 1`] = `
"<form class=\\"has-success\\">
                                    <input required=\\"\\" value=\\"x\\" class=\\"has-success\\">
                                </form>"
`;

exports[`error messages default should replace existing error message with new 1`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" maxlength=\\"3\\" class=\\"has-error\\"><p class=\\"form-error\\">This field must not exceed %s characters in length.</p>
                                </form>"
`;

exports[`error messages default should replace existing error message with new 2`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" maxlength=\\"3\\" class=\\"has-error\\"><p class=\\"form-error\\">This field must not exceed %s characters in length.</p>
                                </form>"
`;

exports[`error messages grouped should display error messages grouped at the bottom (default) 1`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" class=\\"has-error\\">
                                    <input required=\\"\\" minlength=\\"2\\" value=\\"x\\" class=\\"has-error\\">
                                    <ul class=\\"form-errors\\">
                                        <li>This field is required.</li>
                                        <li>This field must be at least %s characters in length.</li>
                                    </ul>
                                </form>"
`;

exports[`error messages grouped should display error messages grouped at the top 1`] = `
"<form class=\\"has-error\\">
                                    <ul class=\\"form-errors\\">
                                        <li>This field is required.</li>
                                        <li>This field must be at least %s characters in length.</li>
                                    </ul>
                                    <input required=\\"\\" class=\\"has-error\\">
                                    <input required=\\"\\" minlength=\\"2\\" value=\\"x\\" class=\\"has-error\\">
                                </form>"
`;

exports[`error messages grouped should hide existing group error message if group is now valid 1`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" class=\\"has-error\\">
                                    <input required=\\"\\" minlength=\\"2\\" class=\\"has-error\\">
                                    <ul class=\\"form-errors\\">
                                        <li>This field is required.</li>
                                        <li>This field must be at least %s characters in length.</li>
                                    </ul>
                                </form>"
`;

exports[`error messages grouped should hide existing group error message if group is now valid 2`] = `
"<form class=\\"has-success\\">
                                    <input required=\\"\\" class=\\"has-success\\">
                                    <input required=\\"\\" minlength=\\"2\\" class=\\"has-success\\">
                                    <ul class=\\"form-errors is-hidden\\"></ul>
                                </form>"
`;

exports[`error messages grouped should not display error messages on valid form 1`] = `
"<form class=\\"has-success\\">
                                    <input required=\\"\\" value=\\"x\\" class=\\"has-success\\">
                                    <input required=\\"\\" minlength=\\"2\\" value=\\"xx\\" class=\\"has-success\\">
                                </form>"
`;

exports[`error messages grouped should replace existing group error messages 1`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" class=\\"has-error\\"><p class=\\"form-error\\">This field is required.</p>
                                    <input required=\\"\\" minlength=\\"2\\" class=\\"has-error\\"><p class=\\"form-error\\">This field must be at least %s characters in length.</p>
                                </form>"
`;

exports[`error messages grouped should replace existing group error messages 2`] = `
"<form class=\\"has-error\\">
                                    <input required=\\"\\" class=\\"has-success\\">
                                    <input required=\\"\\" minlength=\\"2\\" class=\\"has-error\\">
                                    <ul class=\\"form-errors\\">
                                        <li>This field must be at least %s characters in length.</li>
                                    </ul>
                                </form>"
`;
